backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/comments/CommentsService.java

```diff
@@ -22,37 +22,104 @@
 
 package de.metas.ui.web.comments;
 
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import de.metas.comments.CommentEntry;
 import de.metas.comments.CommentEntryRepository;
 import de.metas.ui.web.comments.json.JSONComment;
 import de.metas.ui.web.comments.json.JSONCommentCreateRequest;
+import de.metas.ui.web.view.IViewRow;
+import de.metas.ui.web.window.datatypes.DocumentPath;
 import de.metas.ui.web.window.datatypes.json.DateTimeConverters;
+import de.metas.ui.web.window.descriptor.factory.DocumentDescriptorFactory;
 import de.metas.user.api.IUserDAO;
 import de.metas.util.GuavaCollectors;
+import de.metas.util.ImmutableMapEntry;
 import de.metas.util.Services;
 import lombok.NonNull;
 import org.adempiere.util.lang.impl.TableRecordReference;
 import org.springframework.stereotype.Service;
 
+import javax.annotation.Nullable;
 import java.time.ZoneId;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.Comparator;
+import java.util.IdentityHashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
 
 @Service
 public class CommentsService
 {
 	private final CommentEntryRepository commentEntryRepository;
+	private final DocumentDescriptorFactory documentDescriptorFactory;
 	final IUserDAO userDAO = Services.get(IUserDAO.class);
 
-	public CommentsService(final CommentEntryRepository commentEntryRepository)
+	public static final IdentityHashMap<IViewRow, Boolean> NO_COMMENTS = new IdentityHashMap<>(0);
+
+	public CommentsService(final CommentEntryRepository commentEntryRepository, final DocumentDescriptorFactory documentDescriptorFactory)
 	{
 		this.commentEntryRepository = commentEntryRepository;
+		this.documentDescriptorFactory = documentDescriptorFactory;
+	}
+
+	@NonNull
+	public final IdentityHashMap<IViewRow, Boolean> hasComments(@Nullable final IViewRow row)
```

Comment:
why not `@NonNull row`?

Reply:
at some point i got an error in a view, not sure where.